code/modules/food_and_drinks/kitchen_machinery/kitchen_machine.dm

```diff
@@ -288,11 +288,15 @@
 
 			qdel(O)
 		source.reagents.clear_reagents()
-
-		var/reagents_per_serving = temp_reagents.total_volume / efficiency
-		for(var/i in 1 to efficiency) // Extra servings when upgraded, ingredient reagents split equally
+		var/portions = recipe.duplicate ? efficiency : 1
+		var/reagents_per_serving = temp_reagents.total_volume / portions
+		if(recipe.duplicate)
+			for(var/i in 1 to portions) // Extra servings when upgraded, ingredient reagents split equally
+				var/obj/cooked = new recipe.result(loc)
+				temp_reagents.trans_to(cooked, reagents_per_serving, no_react = TRUE) // Don't react with the abstract holder please
+		else
 			var/obj/cooked = new recipe.result(loc)
-			temp_reagents.trans_to(cooked, reagents_per_serving, no_react = TRUE) // Don't react with the abstract holder please
+			temp_reagents.trans_to(cooked, temp_reagents.total_volume, no_react = TRUE) // Don't react with the abstract holder please
```

Comment:
I don't think you need to branch on `if(recipe.duplicate)` here. If we aren't duplicating, `reagents_per_serving` should be `total_volume / 1` = `total_volume` anyway, and `for(var/i in 1 to 1)` should still run the loop body once.

Reply:
correct, I forgot to change that when Snowball helped me cut down on the code here. Thanks